# Lab03 - TABULAR DATA MANIPULATION

## Part 1:
**Read file “Fruit Dataset.txt” into DataFrame and perform the following tasks:**

Hint read file txt : ```Data = pd.read_csv('data/Fruit Dataset.txt', sep='\t')```
- Display 10 first lines of dataframe
- Show size of DataFrame (number of lines and number of column)
- Display column's name of DataFrame in List type.
- Compute the statistical values: min, max, mean, median, standard deviation, Q1, Q2, and Q3 for all numerical columns in the DataFrame (Hint: use pandas.DataFrame.describe).
- Extract the rows containing the maximum and minimum values of mass, width, height, and color_score.
- Extract the rows where mass falls within the interquartile range (Q1 to Q3).
- Determine the minimum and maximum color_score values for each fruit type.
- Compute the mean and standard deviation of color_score for each fruit type.
- Create a DataFrame with the columns mass, width, height, and color_score positioned at the beginning. Next, include a numerically encoded fruit_subtype column, followed by fruit_name as the last column, to prepare for the next steps.

### Library

In [747]:
import pandas as pd
import numpy as np

### Input file

In [749]:
fruit_data = pd.read_csv('Fruit Dataset.txt', sep='\t')

In [750]:
fruit_data

fruit_label fruit_name     fruit_subtype  mass  width  height  color_score
0             1      apple      granny_smith   192    8.4     7.3         0.55
1             1      apple      granny_smith   180    8.0     6.8         0.59
2             1      apple      granny_smith   176    7.4     7.2         0.60
3             2   mandarin          mandarin    86    6.2     4.7         0.80
4             2   mandarin          mandarin    84    6.0     4.6         0.79
5             2   mandarin          mandarin    80    5.8     4.3         0.77
6             2   mandarin          mandarin    80    5.9     4.3         0.81
7             2   mandarin          mandarin    76    5.8     4.0         0.81
8             1      apple          braeburn   178    7.1     7.8         0.92
9             1      apple          braeburn   172    7.4     7.0         0.89
10            1      apple          braeburn   166    6.9     7.3         0.93
11            1      apple          braeburn   172    7.1     7.6         0.92
12            1      apple          braeburn   154    7.0     7.1         0.88
13            1      apple  golden_delicious   164    7.3     7.7         0.70
14            1      apple  golden_delicious   152    7.6     7.3         0.69
15            1      apple  golden_delicious   156    7.7     7.1         0.69
16            1      apple  golden_delicious   156    7.6     7.5         0.67
17            1      apple  golden_delicious   168    7.5     7.6         0.73
18            1      apple       cripps_pink   162    7.5     7.1         0.83
19            1      apple       cripps_pink   162    7.4     7.2         0.85
20            1      apple       cripps_pink   160    7.5     7.5         0.86
21            1      apple       cripps_pink   156    7.4     7.4         0.84
22            1      apple       cripps_pink   140    7.3     7.1         0.87
23            1      apple       cripps_pink   170    7.6     7.9         0.88
24            3     orange     spanish_jumbo   342    9.0     9.4         0.75
25            3     orange     spanish_jumbo   356    9.2     9.2         0.75
26            3     orange     spanish_jumbo   362    9.6     9.2         0.74
27            3     orange  selected_seconds   204    7.5     9.2         0.77
28            3     orange  selected_seconds   140    6.7     7.1         0.72
29            3     orange  selected_seconds   160    7.0     7.4         0.81
30            3     orange  selected_seconds   158    7.1     7.5         0.79
31            3     orange  selected_seconds   210    7.8     8.0         0.82
32            3     orange  selected_seconds   164    7.2     7.0         0.80
33            3     orange      turkey_navel   190    7.5     8.1         0.74
34            3     orange      turkey_navel   142    7.6     7.8         0.75
35            3     orange      turkey_navel   150    7.1     7.9         0.75
36            3     orange      turkey_navel   160    7.1     7.6         0.76
37            3     orange      turkey_navel   154    7.3     7.3         0.79
38            3     orange      turkey_navel   158    7.2     7.8         0.77
39            3     orange      turkey_navel   144    6.8     7.4         0.75
40            3     orange      turkey_navel   154    7.1     7.5         0.78
41            3     orange      turkey_navel   180    7.6     8.2         0.79
42            3     orange      turkey_navel   154    7.2     7.2         0.82
43            4      lemon    spanish_belsan   194    7.2    10.3         0.70
44            4      lemon    spanish_belsan   200    7.3    10.5         0.72
45            4      lemon    spanish_belsan   186    7.2     9.2         0.72
46            4      lemon    spanish_belsan   216    7.3    10.2         0.71
47            4      lemon    spanish_belsan   196    7.3     9.7         0.72
48            4      lemon    spanish_belsan   174    7.3    10.1         0.72
49            4      lemon           unknown   132    

### 1.1 Overview data:
Display 10 first lines of DataFrame

In [752]:
fruit_data.head(10)

fruit_label fruit_name fruit_subtype  mass  width  height  color_score
0            1      apple  granny_smith   192    8.4     7.3         0.55
1            1      apple  granny_smith   180    8.0     6.8         0.59
2            1      apple  granny_smith   176    7.4     7.2         0.60
3            2   mandarin      mandarin    86    6.2     4.7         0.80
4            2   mandarin      mandarin    84    6.0     4.6         0.79
5            2   mandarin      mandarin    80    5.8     4.3         0.77
6            2   mandarin      mandarin    80    5.9     4.3         0.81
7            2   mandarin      mandarin    76    5.8     4.0         0.81
8            1      apple      braeburn   178    7.1     7.8         0.92
9            1      apple      braeburn   172    7.4     7.0         0.89

### 1.2. DataFrame Size
Show size of DataFrame (number of lines and number of column)

In [754]:
fruit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59 entries, 0 to 58
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   fruit_label    59 non-null     int64  
 1   fruit_name     59 non-null     object 
 2   fruit_subtype  59 non-null     object 
 3   mass           59 non-null     int64  
 4   width          59 non-null     float64
 5   height         59 non-null     float64
 6   color_score    59 non-null     float64
dtypes: float64(3), int64(2), object(2)
memory usage: 3.4+ KB


In [755]:
fruit_data.shape

(59, 7)

**Dataframe size:** 59 rows and 7 columns.

### 1.3. DataFrame Columns
Display column's name of DataFrame in List type.

In [758]:
fruit_data.columns

Index(['fruit_label', 'fruit_name', 'fruit_subtype', 'mass', 'width', 'height',
       'color_score'],
      dtype='object')

In [759]:
fruit_data.columns.tolist()

['fruit_label',
 'fruit_name',
 'fruit_subtype',
 'mass',
 'width',
 'height',
 'color_score']

### 1.4. Statistic Value
Compute the statistical values: min, max, mean, median, standard deviation, Q1, Q2, and Q3 for all numerical columns in the DataFrame (Hint: use pandas.DataFrame.describe).

In [761]:
fruit_data.describe().T

count        mean        std    min     25%     50%     75%  \
fruit_label   59.0    2.542373   1.208048   1.00    1.00    3.00    4.00   
mass          59.0  163.118644  55.018832  76.00  140.00  158.00  177.00   
width         59.0    7.105085   0.816938   5.80    6.60    7.20    7.50   
height        59.0    7.693220   1.361017   4.00    7.20    7.60    8.20   
color_score   59.0    0.762881   0.076857   0.55    0.72    0.75    0.81   

                max  
fruit_label    4.00  
mass         362.00  
width          9.60  
height        10.50  
color_score    0.93

In the statistical summary table:

* **mean**: The average of the data.
* **std**: Also known as **standard deviation**, it represents the dispersion of the data.
* **min**: The smallest value in the dataset.
* **25%**: Corresponds to **Q1** (the first quartile).
* **50%**: Corresponds to both median and **Q2** (the second quartile).
* **75%**: Corresponds to **Q3** (the third quartile).
* **max**: The largest value in the dataset.


### 1.5. Filter data with conditions

5. Extract the rows containing the maximum and minimum values of `mass`, `width`, `height`, and `color_score`.
   * Find max and min values of variables, then save it.
   * Filter the values in the DataFrame that match the previously identified min and max values.

In [765]:
# Find max values of `mass`, `width`, `height`, `color_score`
max = fruit_data[['mass', 'width', 'height', 'color_score']].max()
max

mass           362.00
width            9.60
height          10.50
color_score      0.93
dtype: float64

In [766]:
# Find min values of `mass`, `width`, `height`, `color_score`
min = fruit_data[['mass', 'width', 'height', 'color_score']].min()
min

mass           76.00
width           5.80
height          4.00
color_score     0.55
dtype: float64

In [767]:
# Rows containing the max and min values
result = fruit_data[(fruit_data['mass'].isin([max['mass'], min['mass']])) | 
                    (fruit_data['width'].isin([max['width'], min['width']])) | 
                    (fruit_data['height'].isin([max['height'], min['height']])) | 
                    (fruit_data['color_score'].isin([max['color_score'], min['color_score']]))]

result.sort_values(by=['mass', 'width', 'height', 'color_score'])

fruit_label fruit_name   fruit_subtype  mass  width  height  color_score
7             2   mandarin        mandarin    76    5.8     4.0         0.81
5             2   mandarin        mandarin    80    5.8     4.3         0.77
49            4      lemon         unknown   132    5.8     8.7         0.73
10            1      apple        braeburn   166    6.9     7.3         0.93
0             1      apple    granny_smith   192    8.4     7.3         0.55
44            4      lemon  spanish_belsan   200    7.3    10.5         0.72
26            3     orange   spanish_jumbo   362    9.6     9.2         0.74

6. Display rows that have `mass` in range of *Q1* and *Q3*

In [769]:
# Calculate Q1 and Q3 of `mass`
mass_Q1 = fruit_data['mass'].quantile(0.25)
mass_Q3 = fruit_data['mass'].quantile(0.75)

# Mass between Q1 and Q3
fruit_data[(fruit_data['mass'] >= mass_Q1) & (fruit_data['mass'] <= mass_Q3)].sort_values(by=['mass'])

fruit_label fruit_name     fruit_subtype  mass  width  height  color_score
22            1      apple       cripps_pink   140    7.3     7.1         0.87
28            3     orange  selected_seconds   140    6.7     7.1         0.72
34            3     orange      turkey_navel   142    7.6     7.8         0.75
39            3     orange      turkey_navel   144    6.8     7.4         0.75
35            3     orange      turkey_navel   150    7.1     7.9         0.75
14            1      apple  golden_delicious   152    7.6     7.3         0.69
57            4      lemon           unknown   152    6.5     8.5         0.72
42            3     orange      turkey_navel   154    7.2     7.2         0.82
40            3     orange      turkey_navel   154    7.1     7.5         0.78
37            3     orange      turkey_navel   154    7.3     7.3         0.79
12            1      apple          braeburn   154    7.0     7.1         0.88
15            1      apple  golden_delicious   156    7.7     7.1         0.69
16            1      apple  golden_delicious   156    7.6     7.5         0.67
21            1      apple       cripps_pink   156    7.4     7.4         0.84
38            3     orange      turkey_navel   158    7.2     7.8         0.77
30            3     orange  selected_seconds   158    7.1     7.5         0.79
29            3     orange  selected_seconds   160    7.0     7.4         0.81
36            3     orange      turkey_navel   160    7.1     7.6         0.76
20            1      apple       cripps_pink   160    7.5     7.5         0.86
19            1      apple       cripps_pink   162    7.4     7.2         0.85
18            1      apple       cripps_pink   162    7.5     7.1         0.83
13            1      apple  golden_delicious   164    7.3     7.7         0.70
32            3     orange  selected_seconds   164    7.2     7.0         0.80
10            1      apple          braeburn   166    6.9     7.3         0.93
17            1      apple  golden_delicious   168    7.5     7.6         0.73
23            1      apple       cripps_pink   170    7.6     7.9         0.88
11            1      apple          braeburn   172    7.1     7.6         0.92
9             1      apple          braeburn   172    7.4     7.0         0.89
48            4      lemon    spanish_belsan   174    7.3    10.1         0.72
2             1      apple      granny_smith   176    7.4     7.2         0.60

### 1.6. Group by 

7. Determine *min* and *max* of `color_score` values for each fruit type.

In [772]:
fruit_data.groupby('fruit_name').agg(
    color_score_min=('color_score', 'min'),
    color_score_max=('color_score', 'max')
).reset_index()

fruit_name  color_score_min  color_score_max
0      apple             0.55             0.93
1      lemon             0.70             0.74
2   mandarin             0.77             0.81
3     orange             0.72             0.82

8. Compute the *mean* and *standard deviation* of `color_score` for each fruit type.

In [774]:
fruit_data.groupby('fruit_name').agg(
    color_score_mean=('color_score', 'mean'),
    color_score_std=('color_score', 'std')
).reset_index()

fruit_name  color_score_mean  color_score_std
0      apple          0.783684         0.124196
1      lemon          0.718125         0.010468
2   mandarin          0.796000         0.016733
3     orange          0.771053         0.029039

### 1.7. Create DataFrame
Create a DataFrame with the columns `mass`, `width`, `height`, and `color_score` positioned at the beginning. Next, include a numerically encoded `fruit_subtype` column, followed by `fruit_name` as the last column, to prepare for the next steps.

In [776]:
df = fruit_data[['mass', 'width', 'height', 'color_score', 'fruit_subtype', 'fruit_name']]
df.loc[:, 'fruit_subtype'] =  df['fruit_subtype'].astype('category').cat.codes
df

mass  width  height  color_score fruit_subtype fruit_name
0    192    8.4     7.3         0.55             3      apple
1    180    8.0     6.8         0.59             3      apple
2    176    7.4     7.2         0.60             3      apple
3     86    6.2     4.7         0.80             4   mandarin
4     84    6.0     4.6         0.79             4   mandarin
5     80    5.8     4.3         0.77             4   mandarin
6     80    5.9     4.3         0.81             4   mandarin
7     76    5.8     4.0         0.81             4   mandarin
8    178    7.1     7.8         0.92             0      apple
9    172    7.4     7.0         0.89             0      apple
10   166    6.9     7.3         0.93             0      apple
11   172    7.1     7.6         0.92             0      apple
12   154    7.0     7.1         0.88             0      apple
13   164    7.3     7.7         0.70             2      apple
14   152    7.6     7.3         0.69             2      apple
15   156    7.7     7.1         0.69             2      apple
16   156    7.6     7.5         0.67             2      apple
17   168    7.5     7.6         0.73             2      apple
18   162    7.5     7.1         0.83             1      apple
19   162    7.4     7.2         0.85             1      apple
20   160    7.5     7.5         0.86             1      apple
21   156    7.4     7.4         0.84             1      apple
22   140    7.3     7.1         0.87             1      apple
23   170    7.6     7.9         0.88             1      apple
24   342    9.0     9.4         0.75             7     orange
25   356    9.2     9.2         0.75             7     orange
26   362    9.6     9.2         0.74             7     orange
27   204    7.5     9.2         0.77             5     orange
28   140    6.7     7.1         0.72             5     orange
29   160    7.0     7.4         0.81             5     orange
30   158    7.1     7.5         0.79             5     orange
31   210    7.8     8.0         0.82             5     orange
32   164    7.2     7.0         0.80             5     orange
33   190    7.5     8.1         0.74             8     orange
34   142    7.6     7.8         0.75             8     orange
35   150    7.1     7.9         0.75             8     orange
36   160    7.1     7.6         0.76             8     orange
37   154    7.3     7.3         0.79             8     orange
38   158    7.2     7.8         0.77             8     orange
39   144    6.8     7.4         0.75             8     orange
40   154    7.1     7.5         0.78             8     orange
41   180    7.6     8.2         0.79             8     orange
42   154    7.2     7.2         0.82             8     orange
43   194    7.2    10.3         0.70             6      lemon
44   200    7.3    10.5         0.72             6      lemon
45   186    7.2     9.2         0.72             6      lemon
46   216    7.3    10.2         0.71             6      lemon
47   196    7.3     9.7         0.72             6      lemon
48   174    7.3    10.1         0.72             6      lemon
49   132    5.8     8.7         0.73             9      lemon
50   130    6.0     8.2         0.71             9      lemon
51   116    6.0     7.5         0.72             9      lemon
52   118    5.9     8.0         0.72             9      lemon
53   120    6.0     8.4         0.74             9      lemon
54   116    6.1     8.5         0.71             9      lemon
55   116    6.3     7.7         0.72             9      lemon
56   116    5.9     8.1         0.73             9      lemon
57   152    6.5     8.5         0.72             9      lemon
58   118    6.1     8.1         0.70             9      lemon

## Part 2: NAIVE BAYES CLASSIFICATION (Bonus)
**Read the file "Fruit Dataset.txt" into a DataFrame and perform the following tasks:**

- Create a DataFrame with the columns mass, width, height, and color_score positioned at the beginning. Next, include a numerically encoded fruit_subtype column, followed by fruit_name as the last column.
- Select any two fruit types and save them into a new DataFrame.
- Split the new DataFrame into training and testing sets, with 30% of the data allocated for testing (Hint: sklearn.model_selection.train_test_split).
- Apply the Naïve Bayes classifier to classify the two fruit types and report the accuracy.
(https://scikit-learn.org/1.5/modules/naive_bayes.html)
- Use the original DataFrame containing all four fruit types, splitting it into a training set (70%) and a testing set (30%).
- Use MultinomialNB to classify all four fruit types and report the accuracy for each fruit type. (https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html)

### 2.1. Create Sample
Choose 2 types of fruit into new dataframe

In [779]:
sample_1 = df[df['fruit_name'].isin(['apple', 'lemon'])].reset_index(drop=True)

In [780]:
sample_1['fruit_name'].value_counts()

fruit_name
apple    19
lemon    16
Name: count, dtype: int64

### 2.2. Train Test Split
Split the new DataFrame into training and testing sets, with 30% of the data allocated for testing (Hint: sklearn.model_selection.train_test_split).

In [782]:
from sklearn.model_selection import train_test_split

In [783]:
X = sample_1.drop(columns='fruit_name')
y = sample_1['fruit_name']

In [784]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

### 2.3. Naive Bayes
Apply the **Naïve Bayes** classifier to classify the two fruit types and report the accuracy. (https://scikit-learn.org/1.5/modules/naive_bayes.html)

In [786]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report

In [787]:
GaussianNB_model = GaussianNB()

In [788]:
GaussianNB_model.fit(X_train, y_train)

GaussianNB()

In [789]:
y_pred = GaussianNB_model.predict(X_test)

In [790]:
print("Number of mislabeled points out of a total %d points: %d" % (X_test.shape[0], (y_test != y_pred).sum()))

Number of mislabeled points out of a total 11 points: 0


In [791]:
# accuracy score of model
accuracy_score(y_pred, y_test)

1.0

### 2.4. MultinomialNB
* Use the original DataFrame containing all four fruit types, splitting it into a training set (70%) and a testing set (30%).
* Use **MultinomialNB** to classify all four fruit types and report the accuracy for each fruit type. (https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html)

In [793]:
X = df.drop(columns='fruit_name')
y = df['fruit_name']

In [794]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [795]:
from sklearn.naive_bayes import MultinomialNB

In [796]:
MultinomialNB_model = MultinomialNB()

In [797]:
MultinomialNB_model.fit(X_train, y_train)

MultinomialNB()

In [798]:
y_pred = MultinomialNB_model.predict(X_test)

In [799]:
print("Number of mislabeled points out of a total %d points: %d" % (X_test.shape[0], (y_test != y_pred).sum()))

Number of mislabeled points out of a total 18 points: 7


In [800]:
# overall accurancy
accuracy_score = accuracy_score(y_test, y_pred)

In [801]:
result = pd.DataFrame({'actual': y_test, 'predict': y_pred})
result

actual predict
0      apple   apple
5   mandarin   lemon
34    orange   lemon
13     apple   apple
45     lemon  orange
53     lemon   lemon
57     lemon   lemon
25    orange  orange
47     lemon  orange
12     apple   apple
49     lemon   lemon
3   mandarin   lemon
36    orange  orange
31    orange  orange
8      apple   apple
17     apple   apple
6   mandarin   lemon
4   mandarin   lemon

In [802]:
accuracy_score

0.6111111111111112

In [803]:
report = classification_report(y_pred, y_test, zero_division=0)

In [804]:
print(report)

              precision    recall  f1-score   support

       apple       1.00      1.00      1.00         5
       lemon       0.60      0.38      0.46         8
    mandarin       0.00      0.00      0.00         0
      orange       0.75      0.60      0.67         5

    accuracy                           0.61        18
   macro avg       0.59      0.49      0.53        18
weighted avg       0.75      0.61      0.67        18



**precision** shows the accurrancy of each fruit type.

Therefore, the accurancy of each fruit type is: **apple:** 1.00, **lemon:** 0.60, **madarin:** 0.00, **orange:** 0.75.